File: server/mc/1.21.1-neoforge/src/main/java/calebxzhou/rdi/mc/server/rcmd/RcmdCommandSourceStackSource.java
package calebxzhou.rdi.mc.server.rcmd;

import calebxzhou.rdi.mc.rcmd.RcmdSource;
import net.minecraft.commands.CommandSourceStack;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerPlayer;

import java.util.UUID;

public final class RcmdCommandSourceStackSource implements RcmdSource {
    private final CommandSourceStack source;

    public RcmdCommandSourceStackSource(CommandSourceStack source) {
        this.source = source;
    }

    public ServerPlayer getPlayer() {
        return source.getPlayer();
    }

    @Override
    public String name() {
        return source.getTextName();
    }

    @Override
    public UUID playerId() {
        ServerPlayer player = getPlayer();
        return player == null ? RcmdSource.NO_PLAYER_ID : player.getUUID();
    }

    @Override
    public boolean hasPermission(String permission) {
        return source.hasPermission(4);
    }

    @Override
    public void sendFeedback(String message) {
        source.sendSuccess(() -> Component.literal(message), false);
    }

    @Override
    public void sendError(String message) {
        source.sendFailure(Component.literal("[rcmd] " + message));
    }
}

File: server/mc/1.21.1-neoforge/src/main/kotlin/calebxzhou/rdi/mc/server/rcmd/RcmdCommandSourceStackSource.kt
package calebxzhou.rdi.mc.server.rcmd

import calebxzhou.rdi.mc.rcmd.RcmdSource
import net.minecraft.commands.CommandSourceStack
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerPlayer
import java.util.*
import java.util.function.Supplier

class RcmdCommandSourceStackSource(private val source: CommandSourceStack) : RcmdSource {
    val player: ServerPlayer?
        get() = source.player

    override fun name(): String {
        return source.textName
    }

    override fun playerId(): UUID {
        val player = this.player
        return if (player == null) RcmdSource.NO_PLAYER_ID else player.getUUID()
    }

    override fun hasPermission(permission: String?): Boolean {
        return source.hasPermission(4)
    }

    override fun sendFeedback(message: String) {
        source.sendSuccess(Supplier { Component.literal(message) }, false)
    }

    override fun sendError(message: String?) {
        source.sendFailure(Component.literal("[rcmd] " + message))
    }
}
